Allow PPI::Statement::Expression in hash key

nyrdz opened the pull request:
Originally, I was playing around with subroutine attributes and encounter an issue with the following code.

Before applying perlimports.
```
package MyPackage;
use Scalar::Util qw( refaddr );

my %attrs;

sub MODIFY_CODE_ATTRIBUTES {
	my ( $package, $subref, @attrs ) = @_;

	#my $addr = refaddr $subref;    # OK ─┐
	#$attrs{$addr} = \@attrs;       # ────┘

	#$attrs{ refaddr($subref) } = \@attrs;    # OK

	$attrs{ refaddr $subref } = \@attrs;    # Doesn't recognize refaddr as a Scalar::Util exported symbol

	return;
}

sub main : FOO {
	print "BAR";
}

1;
```

After applying perlimports.
```
package MyPackage;
use Scalar::Util ();

my %attrs;

sub MODIFY_CODE_ATTRIBUTES {
	my ( $package, $subref, @attrs ) = @_;

	#my $addr = refaddr $subref;    # OK ─┐
	#$attrs{$addr} = \@attrs;       # ────┘

	#$attrs{ refaddr($subref) } = \@attrs;    # OK

	$attrs{ refaddr $subref } = \@attrs;    # Doesn't recognize refaddr as a Scalar::Util exported symbol

	return;
}

sub main : FOO {
	print "BAR";
}

1;
```

oalders commented on lib/App/perlimports/Document.pm:
```diff
@@ -400,12 +401,40 @@ sub _build_possible_imports {
+        if (   $word->statement->isa('PPI::Statement::Expression')
+            && $word->isa('PPI::Token::Word')
+            && $word->content =~ /^[a-zA-Z_][a-zA-Z0-9_]*$/ ) {
+            my @children = $word->statement->children;
```
If we use `$word->statement->schildren;` here rather than `$word->statement->children;`, I think we don't need the `next if $current->isa('PPI::Token::Whitespace');` below.

oalders commented on lib/App/perlimports/Document.pm:
```diff
@@ -400,12 +401,40 @@ sub _build_possible_imports {
+                $word_is_string = 1
+                    if $current->isa('PPI::Token::Operator')
+                    && $current->content eq '=>';
+                last;
```
If we've removed the whitespace above, do we even need the `while` loop here anymore? That may also reduce the complexity score.

oalders reviewed (changes requested):
Thanks so much for this contribution. I had a couple of comments.

nyrdz commented:
Didn't knew perlcritic was able to measure code complexity. I find it impressive.

With this last commit, first thing I did in hopes to reduce code complexity was to refactor all logic related to determining if a word is interpreted as a string, into its own subroutine, which was a success. Also, made it so it can check in any kind of statement (not just `PPI::Statement::Expression`).

Then applied your suggestion of changing `$word->statement->children` to `$word->statement->schildren`.

This change revealed a bug in my previous implementation, for some weird reason I wrote
`last unless refaddr($current) == refaddr($word);`
instead of
`last if refaddr($current) == refaddr($word);`
(we want to stop shifting `@children` when our `$word` is found and I was doing the opposite).

Out of curiosity I removed
```
&& !(
       $word->statement
    && $word->statement->isa('PPI::Statement::Variable')
)
```
to see which tests failed. Found out `t/func-in-var.t` which, in turn, used `test-data/func-in-var.pl` failed.

In specific, this part was the problem `$INC{ my $file = class_to_path $class}`, a PPI::Statement::Variable as a hash key. But that can be solved by just using _is_word_interpreted_as_string; and since it's not a word interpreted as a string it must be either a locally defined symbol or an imported one so we push it as a possible import. I removed the previous condition. All tests still passing at this point.

Then I commented out `is_hash_key` and, as expected, lots and lots of tests failed.
Finally, I kept `_is_word_interpreted_as_string` (as below) to see what happened and surprisingly all tests passed. I have no idea why.
```
if (
    !$isa_symbol
    # && is_hash_key($word)
    && $self->_is_word_interpreted_as_string($word)
) {
```

Anyway, I kept `&& is_hash_key($word)` since I don't have a strong argument for removing it, but wanted to point out the fact that it appears to not be necessary for the tests to pass if `_is_word_interpreted_as_string` is used.

oalders reviewed (commented):
This all looks great! I find it easier to follow now as well.

> Anyway, I kept `&& is_hash_key($word)` since I don't have a strong argument for removing it

I'm happy for you to remove it, if you'd like. Basically, if the tests are passing after something has been removed, I'm ok leaving it out. If it turns out it's needed, we can re-add it later with a test case that proves that this code is actually useful.

Do you want to squash everything down into one commit?

nyrdz commented:
So, even though all tests passed before, this perlimports lint error on lib/App/perlimports/Document.pm popped up.
```
-use Text::Diff      ();
+use Text::Diff qw( diff ); <--- because in one line we had $json->{diff} = ....
```

Turns out there was no test to check for single word hash keys so I added t/hash-unquoted-key.t and improved _word_is_interpreted_as_string based on [https://perldoc.perl.org/perlref#Not-so-symbolic-references](https://perldoc.perl.org/perlref#Not-so-symbolic-references)

oalders commented on t/hash-unquoted-key.t:
```diff
@@ -0,0 +1,25 @@
+is(
+    $e->formatted_ppi_statement,
+    'use HTTP::Status ();',
+    'recognizes is_info as an imported symbol'
```
I think this test name needs to be updated?

nyrdz replied:
You're right

oalders commented on test-data/hash-unquoted-key.pl:
```diff
@@ -0,0 +1,7 @@
+use HTTP::Status ();
+
+my %foo;
+$foo{is_info} = 1;
```
It looks like this line has a trailing newline error.

nyrdz replied:
Fixed

oalders reviewed (changes requested):
Excellent finds. This is so close.

oalders commented:
Thanks for all of the quick iterations, @nyrdz!

nyrdz commented:
Thanks to you too, I learned a lot of good stuff.

El sáb, 19 ago 2023 a la(s) 11:46, Olaf Alders ***@***.***)
escribió:

> Thanks for all of the quick iterations, @nyrdz <https://github.com/nyrdz>!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/perl-ide/App-perlimports/pull/102#issuecomment-1685082810>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD7TEGFZCCJE7VOMWRAGCS3XWECYHANCNFSM6AAAAAA3TXORA4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
